polus-image-registration-plugin/src/image_registration.py

```diff
@@ -0,0 +1,477 @@
+
+def get_tile_by_tile_transformation(reference_image_tiles, moving_image_tiles, scale_matrix): 
+    """
+    This function takes as input a set of tiles from the reference image and moving image and
+    calculates the homography tranformation between respective tiles. The transformation is
+    calculated between scaled down versions of the images, the scale matrix is used to upscale
+    the homography matrixes.
+    
+    Input:
+        reference_image_tiles= list of tiles from the reference image
+        moving_image_tiles= list of tiles from the moving image
+        scale_matrix= used to up_scale the homography matrices 
+        
+    Output:
+        homography_set = list of homography matrixes 
+        
+    """   
+    NUMBER_OF_TILES=4
+    homography_set=[]       
+    for k in range(NUMBER_OF_TILES):
+        # cacluate transformation betwen corresponding tiles
+        _ , homography_matrix=image_transformation(moving_image_tiles[k], reference_image_tiles[k] )
+        # upscale and append the matrix
+        homography_set.append(homography_matrix*scale_matrix)            
+    return homography_set
+
+
+def apply_registration(moving_image_path,Template_image_shape,Rough_Homography_Upscaled, fine_homography_set):
+    
+    """
+    This function transforms an image using the set of transformations (rough and fine) returned
+    by the register_images() function below. 
+    
+    Inputs: 
+        moving_image_path: path to the moving image
+        Template_image_shape: dimensions of the reference image
+        Rough_Homography_Upscaled: rough homography matrix returned by register_images() function
+        fine_homography_set: set of tile by tile transformation matrices returned by register_images(function)
+        
+    Outputs:
+        transformed_moving_image: transformed moving image
+        moving_image_metadata: metadata from the moving image       
+        
+    """
+    
+    logging.basicConfig(format='%(asctime)s - %(name)-8s - %(levelname)-8s - %(message)s',
+                        datefmt='%d-%b-%y %H:%M:%S')
+    logger = logging.getLogger("apply_registration()")
+    logger.setLevel(logging.INFO)
+    
+    # read image using bfio
+    bf = BioReader(moving_image_path)
+    moving_image_metadata=bf.read_metadata()
+    moving_image = bf.read_image(C=[0])    
+    moving_image=moving_image[:,:,0,0,0] 
+    del bf
+    
+    mem = psutil.virtual_memory()
+    logger.info('Memory usage at this stage : {}'.format(mem))  
+            
+    height,width=Template_image_shape
+    
+    # apply rough transformation to the moving image
+    rough_transformed_image=apply_rough_homography(moving_image,Rough_Homography_Upscaled,Template_image_shape)
+    del moving_image
+    # get tiles
+    moving_image_tiles=get_tiles(rough_transformed_image, buffer=True)
+    # free up memory
+    del rough_transformed_image    
+ 
+    
+    # apply tile by tile transformation to the tiles
+    NUMBER_OF_TILES=4
```

Comment:
Throughout the code, the number of tiles for the fine transformations remains constant. In the long run, this will not be sustainable. Just to get this plugin out, I think it is okay for now, but please open up an issue in `polus-plugins` and/or add a comment in a `To Do` section of the README to indicate that this will need to be optimized for scaling. This is probably why there were so many memory issues.

Reply:
A note regarding this has been added in the readme under the 'to do' section. Also, a new issue has been opened as a feature request.